apis/Google.Cloud.Firestore/Google.Cloud.Firestore/ValueComparer.cs

```diff
@@ -0,0 +1,251 @@
+                        : leftChar - rightChar; // Not a segment end, so just compare ordinals
+                }
+            }
+            // Shorter comes before longer
+            return leftReference.Length.CompareTo(rightReference.Length);
+        }
+
+        private int CompareArrays(Value left, Value right)
+        {
+            var leftArray = left.ArrayValue.Values;
+            var rightArray = right.ArrayValue.Values;
+            
+            int size = Math.Min(leftArray.Count, rightArray.Count);
+            for (int i = 0; i < size; i++)
+            {
+                int result = Compare(leftArray[i], rightArray[i]);
+                if (result != 0)
+                {
+                    return result;
+                }
+            }
+            return leftArray.Count.CompareTo(rightArray.Count);
+        }
+
+        private int CompareObjects(Value left, Value right)
+        {
+            // This requires iterating over the keys in the object in order and doing a
+            // deep comparison.
+            var leftMap = new SortedDictionary<string, Value>(left.MapValue.Fields, StringComparer.Ordinal);
+            var rightMap = new SortedDictionary<string, Value>(right.MapValue.Fields, StringComparer.Ordinal);
+
+            using (SortedDictionary<string, Value>.Enumerator leftIterator = leftMap.GetEnumerator(),
+                rightIterator = rightMap.GetEnumerator())
+            {
+                bool leftMoveNext = leftIterator.MoveNext();
+                bool rightMoveNext = rightIterator.MoveNext();
+                while (leftMoveNext && rightMoveNext)
+                {
+                    int result = string.Compare(leftIterator.Current.Key, rightIterator.Current.Key);
+                    if (result != 0)
+                    {
+                        return result;
+                    }
+                    result = Compare(leftIterator.Current.Value, rightIterator.Current.Value);
+                    if (result != 0)
+                    {
+                        return result;
+                    }
+                    leftMoveNext = leftIterator.MoveNext();
+                    rightMoveNext = rightIterator.MoveNext();
+                }
+
+                // Only equal if both iterators are exhausted.
+                // (false.CompareTo(true) returns a negative value, which is what we want.)
+                return leftMoveNext.CompareTo(rightMoveNext);
+            }
+        }
+
+        private int CompareNumbers(Value left, Value right)
+        {
+            if (left.ValueTypeCase == ValueTypeOneofCase.DoubleValue)
+            {
+                if (right.ValueTypeCase == ValueTypeOneofCase.DoubleValue)
+                {
+                    return CompareDoubles(left.DoubleValue, right.DoubleValue);
+                }
+                else
+                {
+                    return CompareDoubles(left.DoubleValue, right.IntegerValue);
```

Comment:
`IntegerValue` is a `long`, so this could lead to precision loss and incorrect ordering I think.

Reply:
I was thinking of a situation where:

```C#
double d = (double)(1L << 53);
left.DoubleValue = d;
right.IntegerValue = (long)d + 1;
```

Here the comparison as-is will return 0, but the IntegerValue really does contain a larger value than the DoubleValue, so it should not return 0.
What do you think?

Reply:
I think we can't tell whether the double value *actually* came from `1 << 53`, or whether it came from `1 << 53 + 1.5` (unless I'm missing something). I don't think we can give any value which will be correct more often than it's not correct, so we should stick with simplicity. (This is what the Java code does, btw - and quite possibly what the server code does too.)

Reply:
OK, sounds reasonable.